### ISSUE TITLE
Caddyfile: provide overrideable defaults for virtual hosts

### ISSUE THREAD

**Issue by ionphractal** [mholt/caddy#1469]
Hi,

I have a feature request regarding virtual host configuration. (Or maybe it's already possible?)

### 1. What version of Caddy are you running (`caddy -version`)?
Caddy 0.9.5 (+ccdc286 Thu Feb 23 14:13:23 UTC 2017)

### 2. What are you trying to do?
Set defaults for all virtual hosts.

### 3. What is your entire Caddyfile?
```text
:80 { 
        log stdout
        errors stderr
        redir https://{host}{uri}
}

:443 {
        log stdout
        errors stderr
        gzip

        tls /caddy/ssl/default.crt /caddy/ssl/default.key

        root /caddy/sites.d/0-default
        markdown /
}

site1.local {
        log stdout
        errors stderr
        gzip 

        proxy / dockerhost.local:32000 {
                policy least_conn
                transparent
                insecure_skip_verify
                fail_timeout 10s
                max_fails 1
                try_duration 1s
        }
}
site2.local {
        log stdout
        errors stderr
        gzip 

        proxy / dockerhost.local:32001 {
                policy least_conn
                transparent
                insecure_skip_verify
                fail_timeout 10s
                max_fails 1
                try_duration 1s
        }
}
siteN.local ...
```

### 4. How did you run Caddy (give the full command and describe the execution environment)?
/go/bin/caddy -conf /caddy/Caddyfile -http2=true -agree -log stdout -pidfile /run/caddy.pid

### 5. What I'd love to do
Especially in the context of f49e0c9b560ea7efc25c0b15d422b59f42a6edb1 / #1464 I'd find it very helpful to be able to configure Caddy-wide defaults that may be overwritten by virtual host configuration.
As far as I understand, you currently have to configure every virtual host on its own.

Given above example, it seems more readable to me to write
```text
defaults { 
        log stdout
        errors stderr
        gzip
        timeouts {
                read 60
                header 30
                write  0
                idle 30
        }
}
proxy_defaults {
        policy least_conn
        transparent
        insecure_skip_verify
        fail_timeout 10s
        max_fails 1
        try_duration 1s
}

:80 {
        redir https://{host}{uri}
}

:443 {
        tls /caddy/ssl/default.crt /caddy/ssl/default.key

        root /caddy/sites.d/0-default
        markdown /
}

site1.local {
        proxy / dockerhost.local:32000
}
site2.local {
        gzip off
        proxy / dockerhost.local:32001 {
                max_fails 10
        }
}
siteN.local ...
```

**Comment by AndreKR** [comment]
@ionphractal I found the following workaround which might be interesting for you:

Have a file called for example common.conf:
```
        log stdout
        errors stderr
        gzip
        timeouts {
                read 60
                header 30
                write  0
                idle 30
        }
```

And then in your Caddyfile:
```
site1.local {
        import common.conf
        proxy / dockerhost.local:32000
}
site2.local {
        import common.conf
        gzip off
        proxy / dockerhost.local:32001 {
                max_fails 10
        }
}
```

**Comment by ionphractal** [comment]
@AndreKR Works great, nice! Thank you!

Now there's only the proxy defaults left... too bad import statement is only supported in the server block and outside of it.

**Comment by mholt** [comment]
There will be _some_ duplication in the Caddyfile, this is expected as there is no inheritance or cascading or variables (in the scripting sense of the word). `import` is definitely the recommended way to go to reduce duplication and maintain clarity.

We may revisit the Caddyfile in the future, but probably not until after v1.0. I'll mark this as deferred so we revisit the issue later. :) Thanks!

**Comment by mholt** [comment]
You can now use Caddyfile [snippets](https://caddyserver.com/docs/caddyfile#snippets) to import shared configuration within the same file. Hope that helps!
